internal/resources/slo/data_source_slo.go

```diff
@@ -0,0 +1,340 @@
+package slo
+
+import (
+	"context"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/grafana/terraform-provider-grafana/internal/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func DatasourceSlo() *schema.Resource {
+	return &schema.Resource{
+		Description: `
+		* [Official documentation](https://grafana.com/docs/grafana-cloud/slo/)
+		* [API documentation](https://grafana.com/docs/grafana-cloud/slo/api/)
+		`,
+		ReadContext: datasourceSloRead,
+		Schema: map[string]*schema.Schema{
```

Comment:
I'm open to feedback here with regards to the Schema for both the DatasourceSLO and the ResourceSLO - it's a bit complex due to the message body/response we get back from our API, but I'm very open to modifying the nesting.

Reply:
Currently, the .tf file expects something along the lines of this. I matched the nesting of the alerting structure to that of our API.
```
resource "grafana_slo_resource" "test" {
  name        = "Terraform Testing"
  description = "Terraform Description"
  query       = "sum(rate(apiserver_request_total{code!=\"500\"}[$__rate_interval])) / sum(rate(apiserver_request_total[$__rate_interval]))"
  objectives {
    objective_value  = 0.995
    objective_window = "30d"
  }
  labels {
    key   = "custom"
    value = "value"
  }
  alerting {
    fastburn {
      annotations {
        key   = "name"
        value = "Critical - SLO Burn Rate Alert"
      }
      annotations {
        key   = "description"
        value = "Error Budget is burning at a rate greater than 14.4x."
      }
      labels {
        key   = "type"
        value = "slo"
      }
    }

    slowburn {
      annotations {
        key   = "name"
        value = "Warning - SLO Burn Rate Alert"
      }
      annotations {
        key   = "description"
        value = "Error Budget is burning at a rate greater than 1x."
      }
      labels {
        key   = "type"
        value = "slo"
      }
    }
  }
}
```

We could consider doing something like this, to remove the nesting. Any thoughts or comments? Relative to the rest of the Grafana resources, ours is a bit complex, and it does currently have a fair bit of nesting.

```
resource "grafana_slo_resource" "test" {
  name        = "Terraform Testing"
  description = "Terraform Description"
  query       = "sum(rate(apiserver_request_total{code!=\"500\"}[$__rate_interval])) / sum(rate(apiserver_request_total[$__rate_interval]))"
  objectives {
    objective_value  = 0.995
    objective_window = "30d"
  }
  labels {
    key   = "custom"
    value = "value"
  }
  fastburn_annotations {
    key   = "name"
    value = "Critical - SLO Burn Rate Alert"
  }

  fastburn_annotations {
    key   = "description"
    value = "Error Budget is burning at a rate greater than 14.4x."
  }
  fastburn_labels {
    key   = "type"
    value = "slo"
  }
  slowburn_annotations {
    key   = "name"
    value = "Warning - SLO Burn Rate Alert"
  }

  slowburn_annotations {
    key   = "description"
    value = "Error Budget is burning at a rate greater than 1x."
  }

  slowburn_labels {
    key   = "description"
    value = "Error Budget is burning at a rate greater than 1x."
  }
}
```

Reply:
Does one allow the terraform author greater flexibility in omitting sections (i.e. if not using alerts)?

I think I prefer the nested one, but I'll admit the (mostly awful) terraform that I've modified had a flatter structure. Not sure whether that implies that flat is more idiomatic terraform?

Reply:
This has been addressed - we will stay with the nested structure.

I've also updated the query structure to be nested as well - shape of the HCL Resource would look like this:

```
resource "grafana_slo" "test" {
  name        = "Complex Resource - Terraform Testing"
  description = "Complex Resource - Terraform Description"
  query {
    query_type = "freeform"
    freeform_query = "sum(rate(apiserver_request_total{code!=\"500\"}[$__rate_interval])) / sum(rate(apiserver_request_total[$__rate_interval]))"
  }
  objectives {
    objective_value  = 0.995
    objective_window = "30d"
  }
  labels {
    key   = "slokey"
    value = "slokey"
  }
  alerting {
    name = "alertingname"

    labels {
      key   = "alertingkey"
      value = "alertingvalue"
    }

    fastburn {
      annotations {
        key   = "name"
        value = "Critical - SLO Burn Rate Alert"
      }
      annotations {
        key   = "description"
        value = "Error Budget is burning at a rate greater than 14.4x."
      }
      labels {
        key   = "type"
        value = "slo"
      }
    }

    slowburn {
      annotations {
        key   = "name"
        value = "Warning - SLO Burn Rate Alert"
      }
      annotations {
        key   = "description"
        value = "Error Budget is burning at a rate greater than 1x."
      }
      labels {
        key   = "type"
        value = "slo"
      }
    }
  }
}
```